match type none gives C error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.4 LTS (WSL 2)
Processor: 16 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0

getwd: /home/daniel/Projects/v
vmodules: /home/daniel/.vmodules
vroot: /home/daniel/Projects/v
vexe: /home/daniel/Projects/v/v
vexe mtime: 2022-07-05 12:20:52
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 8b8667d

Git version: git version 2.25.1
Git vroot status: 0.3-32-g8b8667dd-dirty (15 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg main.v`
```v
module main

fn string_none() string | none {
	return none
}

fn main() {
	x := string_none()
	res := match x {
		string {
			false
		}
		none {
			true
		}
	}
	assert res
}

```

**What did you expect to see?**

it should compile and assert true


**What did you see instead?**
```
==================
/tmp/v_1000/main.16054619987900383251.tmp.c:12109: error: cannot convert 'extern struct IError' to 'void *'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2078,7 +2078,11 @@ fn (mut g Gen) call_cfn_for_casting_expr(fname string, expr ast.Expr, exp_is_ptr
 			g.write('&')
 		}
 	}
-	g.expr(expr)
+	if got_styp == 'none' && !g.cur_fn.return_type.has_flag(.optional) {
+		g.write('(none){EMPTY_STRUCT_INITIALIZATION}')
+	} else {
+		g.expr(expr)
+	}
 	g.write(')'.repeat(rparen_n))
 }
 

--- a/vlib/v/tests/sumtype_with_none_test.v
+++ b/vlib/v/tests/sumtype_with_none_test.v
@@ -1,0 +1,18 @@
+module main
+
+fn string_none() string | none {
+	return none
+}
+
+fn test_sumtype_with_none() {
+	x := string_none()
+	res := match x {
+		string {
+			false
+		}
+		none {
+			true
+		}
+	}
+	assert res
+}

